packages/agreement-process/src/services/readModelService.ts

```diff
@@ -302,109 +305,76 @@ export function readModelServiceBuilder(
-                                descriptorState.suspended,
-                              ],
-                            ],
-                          },
-                        ],
-                      },
-                    },
-                  },
-                },
-              },
-              {
-                $match: {
-                  upgradableDescriptor: { $ne: [] },
-                },
-              },
-            ]
-          : []),
-        {
-          $project: {
-            data: 1,
-            eservices: 1,
-            lowerName: { $toLower: "$eservices.data.name" },
-          },
-        },
-        {
-          $sort: { lowerName: 1 },
-        },
-      ];
+      const agreementFilter = getAgreementsFilters(filters);
 
-      const data = await agreements
-        .aggregate([
-          ...aggregationPipeline,
-          { $skip: offset },
-          { $limit: limit },
-        ])
+      const agreementsData = await agreements
+        .aggregate([agreementFilter])
+        .toArray();
+
+      const eserviceIds = agreementsData.map(
+        (agreement) => agreement.data.eserviceId
+      );
+
+      const eservicesData = await eservices
+        .find({ "data.id": { $in: eserviceIds } })
         .toArray();
 
-      const result = z.array(Agreement).safeParse(data.map((d) => d.data));
+      const eservicesMap = new Map(
+        eservicesData.map((eservice) => [eservice.data.id, eservice.data])
+      );
+
+      const combinedData: Array<{
+        agreement: AgreementReadModel;
+        eservice: EServiceReadModel;
+      }> = agreementsData.flatMap((agreement) => {
+        const eservice = eservicesMap.get(agreement.data.eserviceId);
+        return eservice ? [{ agreement: agreement.data, eservice }] : [];
+      });
+
+      const filteredData = filters.showOnlyUpgradeable
+        ? combinedData.filter((cb) => {
+            const currentDescriptor = cb.eservice.descriptors.find(
+              (descr) => descr.id === cb.agreement.descriptorId
+            );
+            const upgradableDescriptor = cb.eservice.descriptors.filter(
+              (upgradable: DescriptorReadModel) => {
+                // Since the dates are optional, if they are undefined they are set to a very old date
+                const currentPublishedAt =
+                  currentDescriptor?.publishedAt ?? new Date(0);
+                const upgradablePublishedAt =
+                  upgradable.publishedAt ?? new Date(0);
+                return (
+                  upgradablePublishedAt > currentPublishedAt &&
+                  (upgradable.state === descriptorState.published ||
+                    upgradable.state === descriptorState.suspended)
+                );
+              }
+            );
+            return upgradableDescriptor.length > 0;
+          })
+        : combinedData;
+
+      const sortedData = filteredData
+        .slice(offset, offset + limit)
```

Comment:
In this case I applied `limit` and `offset` here instead of in the query, since we are doing the filtering here, in order to have the correct value for `totalCount`